rust/src/protocol_types/credential.rs

```diff
@@ -0,0 +1,89 @@
+use crate::*;
+
+#[derive(
+Debug,
+Clone,
+Hash,
+Eq,
+Ord,
+PartialEq,
+PartialOrd,
+serde::Serialize,
+serde::Deserialize,
+JsonSchema,
+)]
+pub enum CredType {
+    Key(Ed25519KeyHash),
+    Script(ScriptHash),
+}
+
+#[wasm_bindgen]
+#[repr(u8)]
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub enum CredKind {
+    Key,
+    Script,
+}
+
+#[wasm_bindgen]
+#[derive(
+Debug,
+Clone,
+Eq,
+Hash,
+Ord,
+PartialEq,
+PartialOrd,
+serde::Serialize,
+serde::Deserialize,
+JsonSchema,
+)]
+pub struct Credential(pub(crate) CredType);
+
+#[wasm_bindgen]
+impl Credential {
+    pub fn from_keyhash(hash: &Ed25519KeyHash) -> Self {
+        Credential(CredType::Key(hash.clone()))
+    }
+
+    pub fn from_scripthash(hash: &ScriptHash) -> Self {
+        Credential(CredType::Script(hash.clone()))
+    }
+
+    pub fn to_keyhash(&self) -> Option<Ed25519KeyHash> {
+        match &self.0 {
+            CredType::Key(hash) => Some(hash.clone()),
+            CredType::Script(_) => None,
+        }
+    }
+
+    pub fn to_scripthash(&self) -> Option<ScriptHash> {
+        match &self.0 {
+            CredType::Key(_) => None,
+            CredType::Script(hash) => Some(hash.clone()),
+        }
+    }
```

Comment:
We use the pattern `new_variant()` and `as_variant()` in rest of the API.
Here we are using `from_variant()` and `to_variant()`. `from_foo()`, `to_foo()` have different semantic meanings elsewhere. Like `from_json()`/`to_json` or `PlutusData::from_address()`.
See: 
https://github.com/Emurgo/cardano-serialization-lib/blob/f2403651fbe76c95db5e324eb4d62d7765d82501/rust/src/protocol_types/native_script.rs#L35-L126
https://github.com/Emurgo/cardano-serialization-lib/blob/f2403651fbe76c95db5e324eb4d62d7765d82501/rust/src/protocol_types/plutus/plutus_data.rs#L200-L293

Reply:
It’s a legacy namings, and it won’t be changed since it might cause more significant issues during migrations from CSL 11 to 12. However, if you have a good idea on how to unify this naming, you can create a PR or a GitHub issue with a detailed proposal on how we can improve it for everyone. It’s an open-source project, and we would be happy for any contributions.